### ISSUE TITLE
How to make hydrogen use a specific Jupyter console?

### ISSUE THREAD

**Issue by amitnagis** [nteract/hydrogen#533]
I am sorry if its a completely stupid question provided it even makes sense!! I have two different python-2.7s installed on a Mac. The two pythons share the version but differ in certain c libraries. I have two corresponding Jupyter consoles which by default call the two pythons, respectively. How do I make hydrogen bind to a specific Jupyter console  (or equivalently make use of a specific python-2.7 out of the two on my machine)? 

Please let me know if the question needs clarification. I apologize for possibly not using the technical terms like 'console', etc. correctly.

**Comment by rgbkrk** [comment]
> I apologize for possibly not using the technical terms like 'console', etc. correctly.

That's ok, there's a lot of terminology here to get through.

How do you have these separate python setups installed? If you have separate environments, you'll want to create a separate kernel installation for each (so that Hydrogen can pick from them). You can create this from the command line by running:

```
/path/to/python -m pip install ipykernel
/path/to/python -m ipykernel install --user --name SomeNameForEnv
```

This way they're both able to be selected by name.

**Comment by amitnagis** [comment]
> How do you have these separate python setups installed?

Let me give an answer in layman terms: I have two python2.7 , let me call them A and B respectively. A is native python2.7 that is preinstalled with MacOS at /usr/bin/python2.7. B was installed using macports at /opt/local/bin/python2.7. Jupyter was installed twice as follows

`/usr/local/bin/pip install jupyter`
and
`/opt/local/bin/pip-2.7 install jupyter`.

Then, atom was downloaded from the atom website and installed, followed by installation of Hydrogen via atom package manager. Ideally I would like to have the freedom to use Hydrogen with either of /usr/bin/python2.7 and /opt/local/bin/python2.7 but it seems by default it assumes /usr/bin/python2.7.
To give perspective, let me say that when I run `/opt/local/bin/jupyter-2.7 notebook`  in the terminal I work with  /opt/local/bin/python2.7 and access all the installed libraries in the notebook, instead when I run `/usr/local/bin/jupyter notebook`  I work with /usr/bin/python2.7 and can correspondingly call all the installed libraries with this python version.

I apologize for not following your answer. I thought therefore, I should perhaps give a more detailed description of my issue incase I wasn't clear to begin with. It is possible that your aforesaid solution solves my problem- in that case I would be grateful if you could explain your solution in layman terms so I that I can implement it without screwing up existing workflow to access python installations by calling different jupyter notebooks. 

Thanks

**Comment by rgbkrk** [comment]
Thanks for making the write up. It helps for me to see the way in which you want this to work and where the gap is. Sadly, the technical answer to this isn't very pleasing. 😅

## Using `$PATH` to your advantage

Hydrogen will only be able to launch _one_ python by default. It can only run the `python` it sees on your `$PATH`. Hydrogen makes a best faith effort to use your default Python kernel (the runtime) for your interactive Python session. If you want to get things working without the approach I suggested originally, there is another workaround. You will have to run `atom `. explicitly with your `PATH` including `/opt/local/bin/` at the beginning like this:

```
PATH=/opt/local/bin:$PATH atom .
```

You have to fully quit atom (at least on the mac) every time you want to switch to your other python environment though. For this reason, I highly recommend learning about Jupyter's kernel installations and how they can benefit you.

## Kernel specifications (kernel specs)

To make it so your notebook server, nteract, Hydrogen, or any other frontend can run any language runtime (kernel), the concept of a kernel specification was created. These are installed into a known directory. For example, on my mac they're in `~/Library/Jupyter/kernels`:

```
/Users/kylek/Library/Jupyter/kernels/ir:
kernel.js      kernel.json    logo-64x64.png

/Users/kylek/Library/Jupyter/kernels/python2:
kernel.json    logo-32x32.png logo-64x64.png

/Users/kylek/Library/Jupyter/kernels/python3:
kernel.json    logo-32x32.png logo-64x64.png

/Users/kylek/Library/Jupyter/kernels/scala:
kernel.json    launcher.jar   logo-64x64.png
```

Here's what is in my python3 kernel's `kernel.json`:

```json
{
 "argv": [
  "/usr/local/opt/python3/bin/python3.5",
  "-m",
  "ipykernel",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
} 
```

This provides a description for Hydrogen to know how to launch this specific kernel. When I try to run Python code with Hydrogen, I get to choose which runtime I want:

![screen shot 2016-12-26 at 11 39 04 am](https://cloud.githubusercontent.com/assets/836375/21485914/07c66ab8-cb60-11e6-9f5f-c4d5db3f3919.png)

Note that it shows both the Python 2 and Python 3 kernel. This is based on the current language shown inside of Atom as well as the fact that the kernel spec has a `language` key of `python`. What shows in the Atom menu is the `display_name` from the kernel spec.

There's an excellent post from @mottalrd on [installing kernel specs per environment](http://www.alfredo.motta.name/create-isolated-jupyter-ipython-kernels-with-pyenv-and-virtualenv/) (what I would call your two python installations). It hopefully gives another perspective on this that I can't provide, as well as the means to set up these kernel specs.

**Comment by amitnagis** [comment]
Hey! thanks for the elaborate reply! Your first solution worked like a charm- but it was your last post that gave me the understanding and the eventual confidence to implement your first solution. Thanks a lot.
